Backends prototyping

nietras opened the pull request:
Creating a PR for the backends stuff to more easily see the changes.

cc: @mdabros

nietras commented on src/SharpLearning.Backend.Cntk/SharpLearning.Backend.Cntk.csproj:
```diff
@@ -0,0 +1,67 @@
+    </Reference>
+    <Reference Include="System" />
+    <Reference Include="System.Core" />
+    <Reference Include="System.Xml.Linq" />
```
clean up includes. just the core ones should be needed.

nietras commented on src/SharpLearning.Backend.Cntk/packages.config:
```diff
@@ -0,0 +1,4 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="CNTK.GPU" version="2.2.0" targetFramework="net461" />
```
can we support both GPU and CPU by using both packages? would we need two projects for that? Does CNTK not have a separate package independent of device?

mdabros replied:
The CNTK.GPU package supports both CPU and GPU, while the CNTK.CPUOnly, only supports CPU. As far as I know, the CPUOnly package was only made in order to minimize the size of the package, avoiding all the GPU dependencies, for CPU only uses. So we should be able to support both devices only using the GPU package.

This might be different with tensorflow, which to my knowledge has a separate build for CPU and GPU.

nietras replied:
Ok, we would have the same issue then, if we wanted to minimize size and only want run CPU only, it is a bit unfortunate that "runtime" hasn't been separated from "type definitions/API", so only the project that executes the code would depend on the "runtime". Anyway, for now focus is on training so we have to live with it, but we should keep it in mind when defining types.

nietras commented on src/SharpLearning.Backend/TrainingApi.cs:
```diff
@@ -0,0 +1,87 @@
+            //    datasource prefetcher
+
+            //    for iterations
+            //          Tuple<ITensorSymbol> batch = source.NextBatch(batchSize: 64); // load next minibatch
```
We have to be very careful here... symbols cannot be direct representations of loaded data. This is important. Symbols are not actual "instances". TF couples instances with symbols by coupling placeholder symbols to actual instances if I remember correctly.

The same applies to all the below. The suggestion is based on symbols, but it is unclear whether the suggestion is actually intending to use concrete instances or define some kind of operation via symbols.

No doubt this will get more concrete, when we start actually working on it. The notion of forward/backward is interesting, although I have not seen this divide explicitly in TF for example. 

Also, don't like "Tuple<>" use ;)

mdabros replied:
Yes, this is rather superficial, but does serve as a starting point to chose a direction from. When we start to implement types, it will become more concrete and then we can decide which parts should be handled via symbols and which part should be concrete instances.

I agree with the use of Tuple<>, it is just a placeholder until we decide on the real types.

nietras commented:
@mdabros did a bit of updating and cleanup for projects. Tensorflow backend project needs to target net461 since TensorFlowSharp is not net standard compatible. Updated to 1.4.0 too, also updated cntk.

Cntk project should be changed to sdk project type, but for now I am just continuing on trying to get minimal backends in place.

nietras commented:
@mdabros I should note that the way these wrapper libraries for TensorFlow and Cntk have been made makes it annoying to use due to the fixed targetting of x64 and net461. The problem is that the packages are poorly factored, there is no split between api surface (pinvoke) and runtime. Like we have done for some of our native stuff.

nietras commented:
We should consider adding a Backends solution only, for only these projects, otherwise the solution is too heavy to work with, especially with regards to tests. Right now I simply unload all the other projects.

nietras commented on src/SharpLearning.Backend.Cntk.Test/CntkBackendTest.cs:
```diff
@@ -0,0 +1,39 @@
+            using (var b = new CntkBackend(deviceType))
+            using (var g = b.CreateGraph())
+            {
+                g.Placeholder(DataType.Single, new int[] { 1, 2, 3 }, "p");
```
This fails with:
```
Message: Test method SharpLearning.Backend.Cntk.Test.CntkBackendTest.CntkBackendTest_CreateGraph_Placeholder threw exception: 
System.TypeInitializationException: The type initializer for 'CNTK.CNTKLibPINVOKE' threw an exception. ---> System.TypeInitializationException: The type initializer for 'SWIGExceptionHelper' threw an exception. ---> System.DllNotFoundException: Unable to load DLL 'Cntk.Core.CSBinding-2.3.1.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
```
@mdabros have you seen this before? The nuget package has been added to the project...

nietras replied:
The dll can be found at:
```
SharpLearning\packages\CNTK.GPU.2.3.1\support\x64\Release
```
but it is not copied to output.

nietras replied:
The CNTK package definition is really weird... seems to be ill defined.

mdabros replied:
No, I have not experienced that error before. But I do know that there are some issues with the current CNTK package: https://github.com/Microsoft/CNTK/issues/2527 and https://github.com/Microsoft/CNTK/issues/2807

However, I am not sure if that relates to the error above.

mdabros replied:
As discussed, I tried changing the `SharpLearning.Backend.Cntk.Test` test project to .net framework to see if this was causing the issue. Sadly, this still result in the same error.

mdabros replied:
It seems that the only way to get this running is to add a x64 solution platform, and run the test using this specific platform. Using anycpu together with x64 test settings is not enough. 

I guess this ties back to the packaging you mentioned earlier. This might also be another indication that the backend projects should have their own solution.

nietras replied:
@mdabros please do not add x64 to solution platform, that is not necessary.

nietras replied:
@mdabros no doubt this is because the Cntk `targets` file does some logic on the Platform attribute of the project. So basically, you just have to change the Cntk.Test project from AnyCPU to x64, you can try and make a new branch from `backends` and create a PR back to it again. The devil lies in trying not to mess all project files up, which can quickly become a mess ;)

nietras replied:
You don't need to change the Solution platform in order to do this. We might still consider whether we should in fact have just a x64 solution format since a lot is tied into this. But hopefully Cntk will become more cross-plat over time. And TF is.

mdabros replied:
@nietras Good warning, I will change only the Cntk.Test project to x64. On a side note, it seems we still have to change the Cntk.Test project format to .net framework. If not, it is not possible to use CNTK directly from the project. I think this related to the two issues I linked to above.

mdabros commented on src/SharpLearning.Backend.TensorFlow.Test/TensorFlowRawMnistTest.cs:
```diff
@@ -0,0 +1,344 @@
+    }
+
+    // Helper class used to load and work with the Mnist data set
+    public class Mnist
```
@nietras This might not be the ideal way to represent and load the mnist dataset. But I guess it will suffice for initial testing.

nietras replied:
@mdabros nah it is pretty dirty and quick way of loading as I said.

mdabros commented on src/SharpLearning.Backend.TensorFlow.Test/TensorFlowRawMnistTest.cs:
```diff
@@ -17,18 +17,144 @@ public class TensorFlowRawMnistTest
+                // Need to do this in loop, can't do it before we have actual variables...
+                // TODO: Is this how to do it???
+                // https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/apply-gradient-descent
+                var adjustedW = g.ApplyGradientDescent(W, learningRate, backprop);
```
@nietras The tensorflow way, in python, to apply gradient descent is to use the optimizer classes. They are defined here in python: 
[Optimizer](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/optimizer.py), This is the base class.
[GradientDescent](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/gradient_descent.py), the most basic of the concrete optimizer implementations.

As you can see in the GradientDescent implementation, that optimizer uses the `apply_gradient_descent` operator. So this must be the correct operator. The base optimizer contains all the setup steps, where all trainable variables are extracted and gradients calculated before the actual call to ApplyGradientDescent, which takes care of updating the variables, representing the weights of the network.

I have an ongoing effort started here to implement a simple version of gradient descent in TensorFlowSharp:
(GradientDescent)[https://github.com/mdabros/TensorFlowSharp/blob/gradient-descent-optimizer/TensorFlowSharp/Training/SGD.cs]

We will probably need a high level concept like this in TensorFlowSharp, in order to match the optimizer/trainer type in CNTK. This again depends on how we would like the actual learning/training api to look.

nietras commented:
@mdabros could you merge master into this, so my tensorflow PR doesn't have this changes shown?

mdabros commented:
@nietras The master has been merge to the backends branch, with the additional changes since you merged, so you can now merge the backends branch to the tensorflow PR to remove the changes shown.

mdabros commented:
Since work on this has currently stopped, I am closing the pull request. We can always re-open it again if it becomes relevant.
